### ISSUE TITLE
Support for enums?

### ISSUE THREAD

**Issue by faultyserver** [Crecto/crecto#90]
An application that I'm working on has a number of state machines (potentially even multiple per model) that I want to represent with an Enum in Crystal.

Right now, I have my state machines implemented with Strings like this:

```crystal
class Thing < Crecto::Model
  schema "things" do
    field :state, String
  end

  STATES = ["created", "started", "finished"]

  validates_inclusion :state, STATES
end
```

This works, but now if I want to validate the state of the model somewhere later on, I have to do a string comparison:

```crystal
def do_something_only_if_started
  return unless self.state == "started"
  ...
end
```

which is prone to typos and is inherently slow as a String comparison.

This _could_ be implemented with constants to avoid the potential typos, but that doesn't solve the string comparison problem:

```crystal
class Thing < Crecto::Model
  module State
    CREATED  = "created"
    STARTED  = "started"
    FINISHED = "finished"
  end
  
  STATES = [State::CREATED, State::STARTED, State::FINISHED]

  validates_inclusion :state, STATES

  def do_something_only_if_started
    return unless self.state == State::STARTED
    ...
  end
end
```

### With Enums

Using an Enum, however, gets the benefits of typo-avoidance, faster comparisons (Enums default to `Int32`, can be even smaller), _and_ looks cleaner, imo.

```crystal
class Thing < Crecto::Model
  enum State
    CREATED
    STARTED
    FINISHED
  end

  schema "things" do
    field :state, State
  end

  # This wouldn't be necessary
  # validates_inclusion :state, STATES

  def do_something_only_if_started
    return unless self.state == State::STARTED
    ...
  end
end
```

There seems to be a general consensus that [enums directly in databases are a bad practice](http://komlenic.com/244/8-reasons-why-mysqls-enum-data-type-is-evil/) (the validity of which I don't 100% agree with, but whatever). That, and limited support across SQL implementations means that trying to represent enums in the database should probably still be done with strings.

This shouldn't be too difficult to deal with. `Enum` has [`parse`](https://github.com/crystal-lang/crystal/blob/917971dedc08e57f18f794f8d47ffc02a63b6159/src/enum.cr#L411) and [`to_s`](https://github.com/crystal-lang/crystal/blob/917971dedc08e57f18f794f8d47ffc02a63b6159/src/enum.cr#L95) to convert to and from String values, so treating `field` definitions with an Enum type could pretty transparently convert to and from Strings to interact with the database.

### Alternatives

There's the wonderful [aasm.cr](https://github.com/veelenga/aasm.cr), but it is really just a direct copy of the Ruby implementation and uses symbols for the state values. Symbols are certainly better than Strings, but [lack the compile-time benefits of Enums](https://crystal-lang.org/docs/syntax_and_semantics/enum.html#usage).

... I don't know of any other packaged state machine builders for Crystal.

**Comment by faultyserver** [comment]
Looking at the [current implementation of `field`](https://github.com/Crecto/crecto/blob/master/src/crecto/schema.cr#L70), I think trying to cram Enum types in there would add too much complexity to it. Maybe, at least to start, it would be simpler to add a new `enum_field` that is responsible for making the right declarations.

An example use of it would probably look something like this:

```crystal
class Vehicle < Crecto::Model
  enum State
    OFF
    STOPPED
    MOVING
  end

  enum Type
    SEDAN
    TRUCK
  end

  schema "things" do
    field       :model,    String
    field       :capacity, Int32
    enum_field  :state,    State
    enum_field  :type,     Type
  end
end
```

`enum_field` would probably end up declaring an implicit `String` field to send to the database, and maybe create a virtual field on top of that to cast to/from the Enum? That's the best I can think of at this point.

**Comment by fridgerator** [comment]
Support for Enum's would be beneficial.  I agree your solution to use `enum_field` might be better off than extending `field`.

**Comment by faultyserver** [comment]
Cool. I've got some time today that I can use to draft a PR.

**Comment by fridgerator** [comment]
closed by #92
